frontends/p4/simplifyDefUse.cpp

```diff
@@ -558,20 +897,67 @@ class FindUninitialized : public Inspector {
         LOG3("FU Visiting " << table->name);
         auto savePoint = ProgramPoint(context, table);
         currentPoint = savePoint;
+        auto saveHeaderDefsBeforeKey = headerDefs->clone();
         auto key = table->getKey();
         visit(key);
+        auto saveHeaderDefsAfterKey = headerDefs->clone();
+        HeaderDefinitions* finalHeaderDefs = nullptr;
         auto actions = table->getActionList();
         for (auto ale : actions->actionList) {
             BUG_CHECK(ale->expression->is<IR::MethodCallExpression>(),
                       "%1%: unexpected entry in action list", ale);
+            headerDefs = saveHeaderDefsAfterKey->clone();
             visit(ale->expression);
             currentPoint = savePoint;  // restore the current point
                                     // it is modified by the inter-procedural analysis
+            if (finalHeaderDefs) {
+                finalHeaderDefs = finalHeaderDefs->intersect(headerDefs);
+            } else {
+                finalHeaderDefs = headerDefs;
+            }
         }
+        if (finalHeaderDefs) {
+            headerDefs = finalHeaderDefs;
+        }
+        headerDefs->setNotReport(saveHeaderDefsBeforeKey);
         LOG3("FU Returning from " << table->name);
         return false;
     }
 
+    void reportWarningIfInvalidHeader(const IR::Expression* expression, const IR::Type *expr_type) {
+        if (!reportInvalidHeaders)
+            return;
+
+        if (expr_type->is<IR::Type_Header>()) {
+            if (auto member = expression->to<IR::Member>()) {
+                // TODO: header unions
```

Comment:
you may be able to model a header_union as a struct of headers.
also, writing to a union field implicitly invalidates all other fields.
this can be a separate PR.